### PR TITLE
proxyd: Add frontend rate limiting

### PR DESCRIPTION
**Pull request by mslipper**
To give us more flexibiltiy with rate limiting, proxyd now supports rate limiting of client (frontend) requests in addition to upstream (backend) requests. This PR also gives us the ability to exempt certain user agents/origins from rate limiting.

### CODE REVIEW

File: proxyd/integration_tests/util_test.go [hunk 1]
```diff
@@ -45,7 +55,13 @@ func (p *ProxydHTTPClient) SendBatchRPC(reqs ...*proxyd.RPCReq) ([]byte, int, er
 }
 
 func (p *ProxydHTTPClient) SendRequest(body []byte) ([]byte, int, error) {
-	res, err := http.Post(p.url, "application/json", bytes.NewReader(body))
+	req, err := http.NewRequest("POST", p.url, bytes.NewReader(body))
+	if err != nil {
+		panic(err)
```
Comment (tynes) [inline proxyd/integration_tests/util_test.go:60]: Are you sure we want to panic here? Could the remote endpoint cause this?

**Review by Inphi — APPROVED** [review]
LGTM

File: proxyd/backend.go [hunk 2]
```diff
@@ -25,8 +25,9 @@ import (
 )
 
 const (
-	JSONRPCVersion       = "2.0"
-	JSONRPCErrorInternal = -32000
+	JSONRPCVersion          = "2.0"
+	JSONRPCErrorInternal    = -32000
+	JSONRPCErrorRateLimited = JSONRPCErrorInternal - 16
```
[the remark this answers is not in the document]
Comment (mslipper) [reply proxyd/backend.go:30]: My concern about doing this is that we'd be modifying a global.

[the remark this answers is not in the document]
Comment (mslipper) [reply proxyd/backend.go]: Updated to clone the error before usage.

**Comment by Inphi** [comment]
yup clone is good. LGTM!

**Pull request merged**